cleverhans/attacks.py

```diff
@@ -48,48 +58,97 @@ def generate(self, x, **kwargs):
                              " provided")
-        return self.sess.run(self._x_adv, feed_dict={self._x: x_val})
+
+        if 'y' in kwargs and kwargs['y'] is None:
+            del kwargs['y']
+
+        # the set of arguments that are structural properties of the attack
+        # if these arguments are different, we must construct a new graph
+        fixed = dict((k, v) for k, v in kwargs.items()
+                     if k not in self.feedable_kwargs)
+
+        # the set of arguments that are passed as placeholders to the graph
+        # on each call, and can change without constructing a new graph
+        feedable = dict((k, v) for k, v in kwargs.items()
+                        if k in self.feedable_kwargs)
+
+        if not all(isinstance(value, collections.Hashable)
+                   for value in fixed.values()):
+            # we have received a fixed value that isn't hashable
+            # this means we can't cache this graph for later use,
+            # and it will have to be discarded later
+            hash_key = None
+        else:
+            # create a unique key for this set of fixed paramaters
+            hash_key = tuple(sorted(fixed.items()))
+
+        if hash_key not in self.graphs:
+            # try our very best to create a TF placeholder for each of the
+            # feedable keyword arguments by inferring the type
+            num_types = (int, float, np.float16, np.float32, np.float64,
+                         np.int8, np.int16, np.int32, np.int32, np.int64,
+                         np.uint8, np.uint16, np.uint32, np.uint64)
+
+            # construct the new keyword arguments as a combination of
+            # (a) the fixed arguments that are graph-specific
+            # (b) the placeholders that correspond to the variable arguments
+            new_kwargs = dict(x for x in fixed.items())
+            for name, value in feedable.items():
+                given_type = self.feedable_kwargs[name]
+                if isinstance(value, np.ndarray):
+                    new_shape = [None]+list(value.shape[1:])
+                    new_kwargs[name] = tf.placeholder(value.dtype, new_shape)
+                elif isinstance(value, num_types):
+                    new_kwargs[name] = tf.placeholder(given_type, shape=[])
+                else:
+                    raise ValueError("Could not identify type of argument " +
+                                     name + ": " + str(value))
+
+            # x is a special placeholder we always want to have
+            x = tf.placeholder(tf.float32, shape=[None]+list(x_val.shape)[1:])
+
+            # now we generate the graph that we want
+            x_adv = self.generate(x, **new_kwargs)
+
+            if hash_key is not None:
+                # only save the graph if every fixed element is hashable
+                self.graphs[hash_key] = (x, new_kwargs, x_adv)
+
+            if len(self.graphs) == 10:
```

Comment:
Maybe `len(self.graphs) >= 10` ?
Also consider graph cache with expiration of old values (maybe in separate pull request after this one)